Update filtering and segmentation lessons

jni opened the pull request:
This updates the filtering and segmentations to be based on the 2021 LMA
workshop
(https://github.com/jni/lma-2021-bioimage-analysis-python/tree/main/lectures).

Motivation:

- filtering has grown more elaborate and now contains a neat little section on
  "the littlest neural network" that students really enjoy.
- a tour of classical segmentation methods is not very relevant these days
  because deep-learning-based segmentation is so much better. Instead I propose
  that we show *one* basic segmentation method so they understand the *idea* of
  a segmentation (and how it's represented in NumPy arrays), and tell them that
  they can swap in any segmentation method they desire. e.g. the
  deep-learning-based Cellpose provides NumPy masks in the exact same format as
  watershed.

jni committed:
Update filtering and segmentation lessons

jni commented:
@lagru this is gonna be the biggest "content" PR I think. Please let me know what you think and whether you want to tinker with something! If you could do the exercises that would be handy too. 😊 I will give myself a refresher tomorrow. Good night! 🙏

lagru committed:
Add missing ipywidgets dependency

required at least for 01_image_filtering.md.

lagru committed:
Remove outdated requirements.txt

environments.yml is the ground truth for which dependencies are needed.
This file can only confuse users and we want everyone to have a
consistent environment anyway and not debug differences between pip and
conda.

lagru committed:
Apply light reformatting

lagru committed:
Update to recommended of NumPy random API

lagru committed:
Use lighter alpha to differentiate line plots

... if there are more than 2.

lagru commented:
Half done going through everything. Seems to have mostly stayed the same since I gave a talk at EuroSciPy. I plan to finish this in the evening.

lagru commented on tutorial/01_image_filtering.md:
```diff
@@ -57,26 +52,27 @@ slideshow:
   slide_type: fragment
 ---
 # Just to make sure we all see the same results
-np.random.seed(0)
```
Unfortunately both APIs give different results for the same "seed". `seed(0)` was actually a really good seed from a teaching perspective because the filters work really well while it still looks random enough. If found `default_rng(44)` to show similar results.

lagru commented on tutorial/01_image_filtering.md:
```diff
@@ -40,8 +37,6 @@ slideshow:
 import matplotlib.pyplot as plt
 import numpy as np
-
-plt.rcParams['image.cmap'] = 'gray'
```
Just curious why you removed this line. To me, grayscale seems the least confusing for new users?

lagru replied:
I am seeing a few places where the cmap is set explicitly (e.g to gray or inferno). Not sure if intentional. I would make it consistent unless you had specific motivations for doing it that way?

jni replied:
yeah, basically I didn't want to modify the matplotlib defaults globally — that has led to other confusing behaviour in the past. "explicit is better than implicit", so now we do one default call (I think?) and then explicitly set the colormap.

Having said this I'm not super-wedded to this!

lagru replied:
> so now we do one default call (I think?) and then explicitly set the colormap.

I am confused what you are suggesting here. If you agree with the premise that grayscale is the way to go, I am happy with going the explicit route and adding `cmap="gray"` everywhere.

jni replied:
Sorry my point is, someone, somewhere, is going to not type `cmap='gray'`, and they will be confused, so it's good to do one call with the default viridis, and go, "this is because image display has a concept of a *colormap*, and this is the default matplotlib colormap for *reasons*, and to make it look normal you do `cmap='gray'`", and then everyone is on the same page from then on. 😊

lagru replied:
:+1:

lagru committed:
Add missing terminator image

lagru commented on tutorial/02_segmentation_and_measuring.md:
```diff
@@ -1,494 +1,598 @@
+markers, _ = ndi.label(mask)
+labels = watershed(-distance, markers, mask=image)
+
+viewer = napari.Viewer()
+image_layer = viewer.add_image(image)
+labels_layer = viewer.add_labels(labels)
+labels_as_image_layer = viewer.add_image(
+    labels, name='labels as image'
+)
```
When running this cell, Napari doesn't display the image fully. Instead part of the viewport is just black. I can move the image around as expected but it's only visible in a part of the viewport:

![image](https://github.com/scipy-2023-image-analysis/tutorial/assets/20140352/bc00ba9d-40d9-445d-8f10-80536838fc67)

Quick searching doesn't immediately give an answer. Have you seen this before? Perhaps related to HiDPI on Linux messing stuff up?

lagru replied:
<strike>Hmm, even stranger. The different layers show beside each other and are not stacked:</strike>

<details>
![image](https://github.com/scipy-2023-image-analysis/tutorial/assets/20140352/92d2dae9-ed97-4f9e-a320-a022c3fa8ca7)
</details>

*Edit: figured out that this issue was because of the grid mode. :sweat_smile: However the issue above still persists. :cry:*

lagru replied:
<details>

```
napari: 0.4.18rc2
Platform: Linux-6.3.9-arch1-1-x86_64-with-glibc2.37
System: Arch Linux
Python: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]
Qt: 5.15.6
PyQt5: 5.15.7
NumPy: 1.25.0
SciPy: 1.11.1
Dask: 2023.6.1
VisPy: 0.12.2
magicgui: 0.7.2
superqt: unknown
in-n-out: 0.1.7
app-model: 0.1.4
npe2: 0.7.0

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 23.1.3
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1707x960, scale 1.5

Settings path:
- /home/lg/.config/napari/image-analysis-23_7f0765db03aaec75ab3cdd8b32a00191e1ef37d1/settings.yaml

```

</details>

lagru replied:
Damn, it works when I set the global scaling in Plasma / X11 back to 100%. So it's a me issue. :pray: Inconvenient for me.

lagru replied:
Wayland to the rescue. :magic_wand: Not perfect but would work if necessary during the tutorial. (Sorry for the noise)

lagru replied:
Found this: https://github.com/napari/napari/issues/1417 :smile: Stéfan seemed to have the exact same issue with cropping.

lagru committed:
Merge branch 'main' into jni-update-lessons

lagru commented on tutorial/02_segmentation_and_measuring.md:
````diff
@@ -1,494 +1,598 @@
 
-### Experimentation: supervised thresholding
+```{code-cell} ipython3
+np.max(cells3d)
````
```suggestion
np.min(cells3d), np.max(cells3d)
```

lagru commented on tutorial/02_segmentation_and_measuring.md:
```diff
@@ -1,494 +1,598 @@
 +++
 
-### Chan-Vese
+Once we have defined our objects, we can make measurements on them using `skimage.measure.regionprops` and the new `skimage.measure.regionprops_table`. These measurements include features such as area or volume, bounding boxes, and intensity statistics.
```
```suggestion
Once we have defined our objects, we can make measurements on them using `skimage.measure.regionprops` and `skimage.measure.regionprops_table`. These measurements include features such as area or volume, bounding boxes, and intensity statistics.
```
Not sure if you still want to call it new as it was added in 0.16?

jni replied:
hahaha very good 😂

lagru commented on environment.yml:
```diff
@@ -12,7 +12,7 @@ dependencies:
   - pooch=1.7.0
   - scikit-learn=1.2.2
-
+  - ipywidgets==8.0.7
```
```suggestion
  - ipywidgets=8.0.7
  - seaborn=0.12.2
```
Used in the [segmentation tutorial](https://github.com/scipy-2023-image-analysis/tutorial/pull/10/files#diff-0ad01c72c8cfde87a9254911db2e1c2a39c22cc9ec5b6c534c4467c3f3750ac8R498).

lagru commented on tutorial/02_segmentation_and_measuring.md:
````diff
@@ -1,494 +1,598 @@
- 
-    return image
-```
+- Read the Cellpose documentation example notebook [here](https://cellpose.readthedocs.io/en/latest/notebook.html).
````
Should cellpose be included in the environment as well?

lagru commented on tutorial/02_segmentation_and_measuring.md:
```diff
@@ -1,494 +1,598 @@
-The data directory also has an excellent image of Stéfan's cat, Chelsea.  With what you've learned, can you segment the cat's nose?  How about the eyes?  Why is the eye particularly challenging?
-
-Hint: the cat's nose is located close to [240, 270] and the right eye center is near [110, 172] in row, column notation.
+Now we register the plugin. **Note: because of a bug, you will need to Kernel -> Restart Kernel if you have created a viewer and closed it in this session.**
```
Is this bug still present? I think it works on my machine? The only warnings that are shown seem to pertain to little-calculator not yet being registered / or registered twice.

jni replied:
I'm not 100%... Maybe we can change "will" to "may".

```suggestion
Now we register the plugin. **Note: because of a bug, you may need to Kernel -> Restart Kernel if you have created a viewer and closed it in this session.**
```

lagru reviewed (commented):
I found the last Napari-based tutorial very interesting! I am updating the plots now and I think then it's ready to merge if my other comments are not touching something critical.

lagru committed:
Add missing dependencies

lagru commented on tutorial/00_images_are_arrays.md:
This cell is currently only showing the 3 images. https://github.com/scipy-2023-image-analysis/tutorial/blob/136cf44df3b465a2d49644c03fba309f422930e3/tutorial/00_images_are_arrays.md?plain=1#L209-L226

I think it's probably fine but we may want to update the explicit `figsize`.

lagru replied:
Done in https://github.com/scipy-2023-image-analysis/tutorial/pull/10/commits/f3ef19909aaeea8d7f1181e45a6dafd184758e02.

lagru committed:
Remove explicit figsize

we don't include as many images as the tutorial we are inheriting from.
The default one is fine to display the 3 images.

lagru commented on tutorial/01_image_filtering.md:
````diff
@@ -532,31 +576,9 @@ def bounded_slice(center, xy_max, size=1, i_min=0):
 ```
 
 ```{code-cell} ipython3
-mean_filter_interactive_demo(bright_square)
````
I can't figure out why the first image in this interactive example is using grayscale. Also weird that the square in the middle isn't white but light gray. Probably not a huge issue but I'd like to be able to explain code I might be talking about. :sweat_smile: 

![image](https://github.com/scipy-2023-image-analysis/tutorial/assets/20140352/0c48fc82-b3e2-478c-9e72-56adfac98152)

jni replied:
I don't know either! This demo was created by Tony Yu for the 2014 (!) tutorial and remains pretty much unchanged since then. 😂 Every year I think "I should update it" and then every year it's the Friday before the tutorial again! 🤣

lagru committed:
Set grayscale colormap explicitly

I do think that for this tutorial gray is the best option because I want
to display actual grayscale images used in this tutorial (camera) in
"true" color. And with that it's easiest to just use gray everywhere.

lagru reviewed (approved):
Feel free to merge if you are happy. You are also welcome to revert https://github.com/scipy-2023-image-analysis/tutorial/pull/10/commits/44cb463fd2e7d111374978d4f5debe3ce997ea1d if you don't agree. :wink:

jni committed:
regionprops_table isn't new anymore!

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

jni committed:
Update tutorial/02_segmentation_and_measuring.md
